### DISCUSSION TITLE
Issue following find_eqpt documentation

### DISCUSSION THREAD

**Discussion by corbinklett** [python-control/python-control#1053] [Q&A] [answered]
I'm trying to use `find_eqpt` in order to trim an aircraft. I need to solve for the trim throttle input and the angle of attack. 

There are a couple of ways that I can't figure out how to use `find_eqpt`, commented below. How do I specify that d/dt of only some of the states (u, w, theta, and q) matters in this calculation, given that other states are specified as fixed? 

Also, it doesn't seem like `find_eqpt` handles bounds or constraints, such that you can constrain a quaternion to norm 1 (this example just uses Euler angles).

```python
# Fix output at:
Vinf = 30
beta = 0
gamma = 0

# Find this output:
alpha = 0 # guess

# Find this input:
throttle = 0.5 # guess

# State guess
x = 0
y = 0
z = 0
u = Vinf
v = 0
w = 0
phi = 0    # Roll angle
theta = 0
psi = 0    # Yaw angle
p = 0
q = 0
r = 0

x0 = np.array([x, y, z, u, v, w, phi, theta, psi, p, q, r])
u0 = np.array([throttle])

trim_state, trim_input, yeq, result = ct.find_eqpt(
  aircraft_6dof_model, 
   x0, 
   u0=u0, 
   y0=[Vinf, alpha, beta, gamma],
   iy=[0, 2, 3],           			# Constrain Vinf, beta, gamma
   ix=[4, 6, 8, 9, 10, 11],   			# Constrain v, phi, psi, p, q, r
   # dx0=[0,0,0,0,0,0,0,0,0,0,0,0],	# How to specify that xdot, ydot, and zdot don't matter?
   idx=[3, 5, 7, 10], 			        # Only care about d/dt of u, w, theta, and q	
   return_y = True,
   return_result = True
)
```
Depending on how I formulate the arguments, I get an input/output mismatch error, or the `find_eqpt` function runs infinitely.

**Comment by murrayrm** [comment]
What you have looks correct.  If you look inside [the code](https://github.com/python-control/python-control/blob/0.10.1/control/nlsys.py#L1666-L1924) at the definition of [`rootfun`](https://github.com/python-control/python-control/blob/6d4208855d95426fe3cf0b000c6adeafea855c1a/control/nlsys.py#L1883), you will see that the call to the root finding code has the following (extracted) lines:
```
dx = sys._rhs(t, x, u) - dx0
dy = sys._out(t, x, u) - y0
return np.concatenate((dx[deriv_vars], dy[output_vars]), axis=0)
```
The `deriv_vars` varaible is `ix` and the `output_vars` variable is `iy`, so only those variables will be constrained to be zero.

I did a run of your above using
```
aircraft_6dof_model = ct.rss(states=12, inputs=1, outputs=4)
```
and it ran OK.

If you post what you are using for `aircraft_6dof_model`, I can see if that is causing the problem.

In terms of including bounds or constraints: that is not supported, but I'm not sure it would help.  The `find_eqpt` will only work if there is a unique (local) equilibrium point and so if you can't find one without the constraints, you won't find one by imposing the constraints.

**Reply by corbinklett** [reply]
Still diagnosing this. Here is the full system code, along with an example trim condition that I found by hand at the bottom. The `find_eqpt` function does not throw an error, but the solver returns a `False` result. I've been thinking that this is may be because the x, y, z position states do not effect the dynamics at all yet seem to get treated as decision variables by the optimizer. I tried removing them from the code and adjusting the `find_eqpt` arguments accordingly, but got another input-output mismatch error. 

```python
import control as ct
import numpy as np

mass = 500

def gravity_model(u):
    phi, theta, psi = u[0:3]  # roll, pitch, yaw
    # Gravity vector in inertial frame (NED)
    g_ned = [0, 0, mass*9.81]
    # Calculate the gravity vector in the body frame using Euler angles
    Fx = -np.sin(theta) * g_ned[2]
    Fy = np.sin(phi) * np.cos(theta) * g_ned[2]
    Fz = np.cos(phi) * np.cos(theta) * g_ned[2]
    return [Fx, Fy, Fz]

def propulsion_model(throttle):
    max_thrust = 1000.0
    return np.array([throttle * max_thrust, 0, 0])

def aerodynamics_model(u):
    density = 1.225
    wing_area = 20.0
    chord_length = 2.0
    span = 10.0
    
    airspeed, alpha, beta, elevator, aileron, rudder = u
    
    # Simplified aerodynamic coefficients
    CL0 = 0
    CL = np.pi * alpha  + CL0 # Lift coefficient
    CD = 0.02 + CL**2 / (np.pi * 0.6 * span / chord_length)  # Drag coefficient
    CY = -0.02 * beta  # Side force coefficient
    
    # Forces
    L = 0.5 * density * airspeed**2 * wing_area * CL
    D = 0.5 * density * airspeed**2 * wing_area * CD
    Y = 0.5 * density * airspeed**2 * wing_area * CY
    
    # Moments (simplified)
    l = 0.1 * aileron  # Roll moment
    m = 0.1 * elevator  # Pitch moment
    n = 0.1 * rudder  # Yaw moment
    
    # Convert to body frame
    Fx = -D * np.cos(alpha) + L * np.sin(alpha)
    Fz = -D * np.sin(alpha) - L * np.cos(alpha)
    Fy = Y
    
    return [Fx, Fy, Fz, l, m, n, L, D]

def dynamics(t, x, u, params):
    pos = x[0:3]  # inertial frame
    vel = x[3:6]  # body frame
    euler = x[6:9]  # [phi, theta, psi] (roll, pitch, yaw)
    omega = x[9:]   # body angular rates [p, q, r]

    inertia = np.array([[600., 0., 0.],
	                    [0., 1000., 0.],
	                    [0., 0., 500.]])
    
    throttle = u[0]

    out = outputs(t, x, u, params)
    airspeed, alpha, beta, gamma = out
    force = propulsion_model(throttle)
    force += gravity_model(euler)
    
    aero = aerodynamics_model(np.array([airspeed, alpha, beta, 0, 0, 0]))
    
    force += np.array([aero[0], aero[1], aero[2]])
    moments = np.array([aero[3], aero[4], aero[5]])
    
    # Create rotation matrix from euler angles
    phi, theta, psi = euler
    c_phi, s_phi = np.cos(phi), np.sin(phi)
    c_theta, s_theta = np.cos(theta), np.sin(theta)
    c_psi, s_psi = np.cos(psi), np.sin(psi)
    
    R_body_to_inertial = np.array([
        [c_theta*c_psi, s_phi*s_theta*c_psi - c_phi*s_psi, c_phi*s_theta*c_psi + s_phi*s_psi],
        [c_theta*s_psi, s_phi*s_theta*s_psi + c_phi*c_psi, c_phi*s_theta*s_psi - s_phi*c_psi],
        [-s_theta, s_phi*c_theta, c_phi*c_theta]
    ])
    
    ddt_pos = R_body_to_inertial @ vel
    ddt_vel = force / mass - np.cross(omega, vel)
    
    # Euler angle rates
    euler_rates_matrix = np.array([
        [1, s_phi*s_theta/c_theta, c_phi*s_theta/c_theta],
        [0, c_phi, -s_phi],
        [0, s_phi/c_theta, c_phi/c_theta]
    ])
    ddt_euler = euler_rates_matrix @ omega
    
    ddt_omega = np.linalg.inv(inertia) @ (moments - np.cross(omega, inertia @ omega))
    
    return np.hstack([ddt_pos, ddt_vel, ddt_euler, ddt_omega])

def outputs(t, x, u, params):
    V_body = x[3:6]
    # assume zero wind
    airspeed = np.linalg.norm(V_body)
    alpha = np.arctan2(V_body[2], V_body[0])

    # Convert body velocity to NED frame using rotation matrix from euler angles
    euler = x[6:9]  # [phi, theta, psi]
    phi, theta, psi = euler
    c_phi, s_phi = np.cos(phi), np.sin(phi)
    c_theta, s_theta = np.cos(theta), np.sin(theta)
    c_psi, s_psi = np.cos(psi), np.sin(psi)
    
    R_body_to_NED = np.array([
        [c_theta*c_psi, s_phi*s_theta*c_psi - c_phi*s_psi, c_phi*s_theta*c_psi + s_phi*s_psi],
        [c_theta*s_psi, s_phi*s_theta*s_psi + c_phi*c_psi, c_phi*s_theta*s_psi - s_phi*c_psi],
        [-s_theta, s_phi*c_theta, c_phi*c_theta]
    ])
    
    V_NED = R_body_to_NED @ V_body
    gamma = -np.arcsin(V_NED[2] / np.linalg.norm(V_NED))

    if airspeed > 0.1:
        beta = np.arcsin(V_body[1] / airspeed)
    else:
        beta = 0
    
    return [airspeed, alpha, beta, gamma]

sys = ct.nlsys(updfcn=dynamics, outfcn=outputs, inputs=1, outputs=4, states=12,name='sys')

Vinf = 30
x0 = np.array([0, 0, 0, Vinf, 0.0, 0.0, 0, 0, 0, 0.0, 0.0, 0.0])
u0 = np.array([0.5]) # throttle only - moment is zero at zero elevator deflection
y0 = np.array([Vinf, 0, 0, 0])

# sys = ct.rss(states=12, inputs=1, outputs=4)
trim_state, trim_input, yeq, result = ct.find_eqpt(
	sys, 
	x0, 
	u0=u0, 
	y0=y0,                    # Include all three outputs
	iy=[0, 2, 3],             # Constrain 3 of 4 outputs
	ix=[4, 6, 8, 9, 10, 11],  # Constrain velocities and angular rates
	idx=[3, 5, 7, 10],
	return_y = True,
	return_result = True
)

# Example trim condition:
alpha = np.radians(8.05)
x0 = np.array([0, 0, 0, Vinf*np.cos(alpha), 0.0, Vinf*np.sin(alpha), 0, alpha, 0, 0.0, 0.0, 0.0])
u0 = np.array([.45])

sys._rhs(0, x0, u0)
sys._out(0, x0, u0)
```

**Reply by murrayrm** [reply]
I put your code into the latest main branch and it returned almost immediately with the following result:
```
In [2]: yeq
Out[2]: array([30.00017353,  0.14058246,  0.        , -0.04797305])
```
This doesn't quite match your hand calculation (which has 0 for the final entry), but it didn't hang and the first three entries match the final calculation that you show:
```
Out[1]: array([ 3.00000000e+01,  1.40499005e-01,  0.00000000e+00, -2.05860969e-17])
```

Can you post the version of Python, python-control, numpy, and scipy that you are using.  For me, they are:
* Python: 3.12.2
* python-control: 0.10.1.dev32+gdbc998de
* numpy: 1.26.4
* scipy: 1.13.0

Also, when I look at `trim_state`, I get:
```
In [5]: trim_state
Out[5]: 
array([-6.93849578e+18, -2.68847397e+17,  0.00000000e+00,  2.97042083e+01,
        0.00000000e+00,  4.20361982e+00,  0.00000000e+00,  9.26094047e-02,
        0.00000000e+00,  0.00000000e+00,  0.00000000e+00,  0.00000000e+00])
```
The large numbers in *x* and *y* may be what is causing the problem.  As you note above, those variables are "free" and so the root finding algorithm in SciPy may be having group converging.

**Reply by corbinklett** [reply]
The x, y large numbers may be why the `result` variable returns:

```
>>> result
 message: The iteration is not making good progress, as measured by the 
            improvement from the last five Jacobian evaluations.
 success: False
  status: 4
     fun: [ 2.526e-04 -2.221e-04  0.000e+00  0.000e+00  1.672e-04
            0.000e+00 -4.750e-02]
       x: [ 6.410e+17  7.337e+17  0.000e+00  2.970e+01  4.203e+00
            9.308e-02  2.180e-01]
  method: hybr
    nfev: 123
    fjac: [[ 1.000e+00  0.000e+00 ...  0.000e+00  0.000e+00]
           [ 0.000e+00  1.000e+00 ...  0.000e+00  0.000e+00]
           ...
           [ 0.000e+00  0.000e+00 ...  0.000e+00 -5.049e-17]
           [ 0.000e+00  0.000e+00 ...  1.000e+00  0.000e+00]]
       r: [-0.000e+00  0.000e+00 ...  0.000e+00 -0.000e+00]
     qtf: [ 2.526e-04 -2.221e-04  0.000e+00  5.682e-05 -4.750e-02
           -1.417e-18  0.000e+00]
```

I get the same exact results that you do, but it's still an incorrect trim result (returns a flight path angle of -2.72 degrees, but the `find_eqpt` `y0`  and `iy` arguments should constrain it to zero).

**Reply by corbinklett** [reply]
Will dig into why I can't set `ix=[0, 1, 2, 4, 6, 8, 9, 10, 11]` in order to fix `x`, `y`, and `z`, at 0.

**Reply by corbinklett** [reply, marked as answer]
The issue is that the problem I desire to solve is over-constrained. I get the intended result if I change the line in `find_eqpt` that calls `root` to `result = root(rootfun, z0, method='lm')`, adding the "least squares method" designator to the args.

Does it make sense for `find_eqpt` to do this if `num_constraints > num_freedoms`?

**Comment by murrayrm** [comment]
One possibility, which we do in other places, is to create a parameter `root_method` that let's you change the method passed to the `root` function.  We could also include `root_kwargs` as a way to pass additional arguments.  See `ct.OptimalControlProblem` for an example of this.

If we add `root_method` and set the default to `None`, then we could also add some logic to check the number of constraints versus degrees of freedom, with the default solver set to 'lm' if `num_constraints > num_freedoms`.  Or just add something to the documentation indicating you should change the method in that case.

It still seems like setting `ix` appropriately should fix the problem.  I'll try looking into that further when I get a chance.

**Reply by murrayrm** [reply]
PR #1054 adds a `root_method` keyword parameter that can be used to fix the issue identified above.
